File: server-core/src/api/tokens.rs
use std::sync::Arc;

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Extension;
use axum::Json;
use serde::{Deserialize, Serialize};

use super::auth::AuthUser;
use super::state::AppState;
use crate::auth;
use crate::db::NewApiToken;

#[derive(Serialize)]
pub struct TokenResponse {
    pub token_id: String,
    pub label: String,
    pub created_at: u64,
    pub expires_at: Option<u64>,
    pub last_used_at: Option<u64>,
}

#[derive(Serialize)]
pub struct CreateTokenResponse {
    pub token_id: String,
    pub raw_token: String,
}

#[derive(Deserialize)]
pub struct CreateTokenRequest {
    pub label: String,
    pub expires_at: Option<u64>,
}

/// GET /api/v1/tokens
pub async fn list(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
) -> Response {
    match state.db.list_api_tokens(&user.user_id).await {
        Ok(tokens) => {
            let responses: Vec<TokenResponse> = tokens
                .into_iter()
                .map(|t| TokenResponse {
                    token_id: t.token_id.0,
                    label: t.label,
                    created_at: t.created_at,
                    expires_at: t.expires_at,
                    last_used_at: t.last_used_at,
                })
                .collect();
            Json(responses).into_response()
        }
        Err(e) => {
            tracing::error!("list tokens error: {e}");
            StatusCode::INTERNAL_SERVER_ERROR.into_response()
        }
    }
}

/// POST /api/v1/tokens
pub async fn create(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Json(body): Json<CreateTokenRequest>,
) -> Response {
    let (token_id, raw_token) = auth::generate_api_token();
    let token_hash = auth::hmac_token(&raw_token, &state.hmac_secret);

    let new_token = NewApiToken {
        token_id: token_id.clone(),
        user_id: user.user_id,
        token_hash,
        label: body.label,
        expires_at: body.expires_at,
    };

    if let Err(e) = state.db.create_api_token(&new_token).await {
        tracing::error!("create token error: {e}");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    Json(CreateTokenResponse {
        token_id: token_id.0,
        raw_token,
    })
    .into_response()
}

/// DELETE /api/v1/tokens/:token_id
pub async fn revoke(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Path(token_id): Path<String>,
) -> Response {
    let tid = ghostcam::types::TokenId(token_id);

    // Verify ownership by checking the token belongs to the user
    match state.db.list_api_tokens(&user.user_id).await {
        Ok(tokens) => {
            if !tokens.iter().any(|t| t.token_id == tid) {
                return StatusCode::NOT_FOUND.into_response();
            }
        }
        Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
    }

    if let Err(e) = state.db.delete_api_token(&tid).await {
        tracing::error!("revoke token error: {e}");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    StatusCode::OK.into_response()
}

File: server/tokens.go
package main

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/cargocam/ghostcam/server/auth"
	"github.com/cargocam/ghostcam/server/db"
	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
)

type tokenResponse struct {
	TokenID    string `json:"token_id"`
	Label      string `json:"label"`
	CreatedAt  int64  `json:"created_at"`
	ExpiresAt  *int64 `json:"expires_at,omitempty"`
	LastUsedAt *int64 `json:"last_used_at,omitempty"`
}

type createTokenRequest struct {
	Label     string `json:"label"`
	ExpiresAt *int64 `json:"expires_at,omitempty"`
}

type createTokenResponse struct {
	TokenID  string `json:"token_id"`
	RawToken string `json:"token"`
}

// ListTokens handles GET /api/v1/tokens.
func (a *App) ListTokens(w http.ResponseWriter, r *http.Request) {
	userID := getUserID(r)

	tokens, err := a.DB.ListAPITokens(r.Context(), userID)
	if err != nil {
		slog.Error("list tokens failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	resp := make([]tokenResponse, 0, len(tokens))
	for _, t := range tokens {
		resp = append(resp, tokenResponse{
			TokenID:    t.TokenID,
			Label:      t.Label,
			CreatedAt:  t.CreatedAt,
			ExpiresAt:  t.ExpiresAt,
			LastUsedAt: t.LastUsedAt,
		})
	}
	writeJSON(w, http.StatusOK, resp)
}

// CreateToken handles POST /api/v1/tokens.
func (a *App) CreateToken(w http.ResponseWriter, r *http.Request) {
	userID := getUserID(r)

	var body createTokenRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if body.Label == "" {
		writeError(w, http.StatusBadRequest, "label is required")
		return
	}

	tokenID := uuid.New().String()
	rawBytes := make([]byte, 32)
	if _, err := rand.Read(rawBytes); err != nil {
		http.Error(w, "", http.StatusInternalServerError)
		return
	}
	rawToken := base64.RawURLEncoding.EncodeToString(rawBytes)
	tokenHash := auth.HMACToken(rawToken, a.HMACSecret)

	err := a.DB.CreateAPIToken(r.Context(), &db.NewAPIToken{
		TokenID:   tokenID,
		UserID:    userID,
		TokenHash: tokenHash,
		Label:     body.Label,
		ExpiresAt: body.ExpiresAt,
	})
	if err != nil {
		slog.Error("create token failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	writeJSON(w, http.StatusCreated, createTokenResponse{
		TokenID:  tokenID,
		RawToken: rawToken,
	})
}

// RevokeToken handles DELETE /api/v1/tokens/{tokenID}.
func (a *App) RevokeToken(w http.ResponseWriter, r *http.Request) {
	tokenID := chi.URLParam(r, "tokenID")
	if err := a.DB.DeleteAPIToken(r.Context(), tokenID); err != nil {
		slog.Error("revoke token failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}
	w.WriteHeader(http.StatusOK)
}
